### PR TITLE
Calypso Component DateRange improvement: Add Shortcuts Menu

### PR DESCRIPTION
**Pull request by annacmc**
## Proposed Changes

This PR adds an optional date range shortcuts menu to the side of the DateRange component. 

## Testing Instructions

**To test component in use in Stats**

* Open live branch
* Navigate to traffic page
* Add flag `stats/date-picker-calendar`
* Open date range picker and try selecting various date ranges, and using the shortcuts menu on the right
* Check on different size displays, browsers, mobile etc. 
* test out that daterange picker works and looks exactly as before when the shortcuts menu is not displayed
* confirm the readme and docs example are updated sufficiently for the component 

image

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Review by kangzj — COMMENTED** [review]
I like where this is heading. Great work @annacmc ! Everything looks great! I think we could try to limit the scope as is and ship it soon 👍

**Comment by annacmc** [comment]
> I like where this is heading. Great work @annacmc ! Everything looks great! I think we could try to limit the scope as is and ship it soon 👍

Awesome, thanks so much @kangzj I really appreciate it! I'm thinking this iteration should be ready to ship once the display shortcut can be toggled on/off and the updated info is added to the storybook/docs. Then leave the others (like passing a custom shortcuts list etc) for a future iteration.

File: client/components/date-range/index.js [hunk 1]
```diff
@@ -578,6 +581,11 @@ export class DateRange extends Component {
 	 * @returns {import('react').Element} the Popover component
 	 */
 	renderPopover() {
+		const { renderShortcuts } = this.props;
+
+		// Only render shortcuts if renderShortcuts is true
+		const shortcuts = renderShortcuts ? this.renderShortcutsComponent() : null;
```
Comment (kangzj) [inline client/components/date-range/index.js]: The line doesn't seem to necessary to me? We could leverage `this.props.renderShortcuts` on L619?

File: client/components/date-range/index.js [hunk 2]
```diff
@@ -578,6 +581,11 @@ export class DateRange extends Component {
 	 * @returns {import('react').Element} the Popover component
 	 */
 	renderPopover() {
+		const { renderShortcuts } = this.props;
```
Comment (kangzj) [inline client/components/date-range/index.js]: the naming is in line with the functions like `renderInputs` but is a flag to identify whether to render short cuts. Maybe we could change it to another name, like `shouldRenderShortcuts`?

Also maybe we could follow the pattern of `renderXxx` to make it more extendible, but it'll probably be another PR 😄

File: client/components/date-range/shortcuts.tsx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	WEEK: 'week',
+	MONTH: 'month',
+};
+
+const DateRangePickerShortcuts = ( {
+	currentShortcut,
+	onClick,
+}: {
+	currentShortcut?: string;
+	onClick: ( shortcut: {
```
Comment (kangzj) [inline client/components/date-range/shortcuts.tsx]: The definition here doesn't match `handleShortcutClick` function in the date range component.

File: client/components/date-range/index.js (hunk 2 above)
Comment (annacmc) [reply client/components/date-range/index.js]: >the naming is in line with the functions like renderInputs but is a flag to identify whether to render short cuts. Maybe we could change it to another name, like shouldRenderShortcuts?

Yes, I think this is a good idea. 

>Also maybe we could follow the pattern of renderXxx to make it more extendible, but it'll probably be another PR

Yes, this is my thinking here. I want to toggle shortcuts on/off with this PR and for it to use the default shortcuts. In the next PR I want to extend it to accept custom shortcuts so we can pass through whatever we want, irrespective to it's re-usability. Like if we have realtime stats or other custom shortcuts.

File: client/components/date-range/index.js (hunk 1 above)
Comment (annacmc) [reply client/components/date-range/index.js]: Yeah, you're right. It's been a bit of a tangle of getting things working, and I've been tidying up and trying to eliminate unnecessary bits and pieces. I'm looking at this part now; thanks for highlighting it!

File: client/components/date-range/shortcuts.tsx [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+				{ shortcutList.map( ( shortcut, idx ) => (
+					<li
+						className={ clsx( 'date-control-picker-shortcuts__shortcut', {
+							'is-selected': shortcut.id === selectedShortcut,
+						} ) }
+						key={ shortcut.id || idx }
+					>
+						<Button
+							onClick={ () => handleClick( shortcut ) }
+							aria-label={ `Time range ${ shortcut.label }. Click to select.` }
```
Comment (getdave) [inline client/components/date-range/shortcuts.tsx]: This string isn't translated.

File: client/components/date-range/shortcuts.tsx [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+						>
+							<span>{ shortcut.label }</span>
```
Comment (getdave) [inline client/components/date-range/shortcuts.tsx:76]: Needs translation.

File: client/components/date-range/shortcuts.tsx [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+}: {
+	currentShortcut?: string;
+	onClick: ( shortcut: { newFromDate: moment.Moment; newToDate: moment.Moment } ) => void;
+} ) => {
+	const translate = useTranslate();
+	const [ selectedShortcut, setSelectedShortcut ] = useState( currentShortcut );
+
+	useEffect( () => {
+		setSelectedShortcut( currentShortcut );
+	}, [ currentShortcut ] );
```
Comment (getdave) [inline client/components/date-range/shortcuts.tsx]: Can you explain why this effect is necessary? 

On each render the `currentShortcut` is set as the initial state in the `useState`. Then the effect will run and do the exact same thing.

File: client/components/date-range/index.js [hunk 7]
```diff
@@ -430,6 +433,11 @@ export class DateRange extends Component {
 	 * @returns {import('react').Element} the Popover component
 	 */
 	renderPopover() {
+		const { displayShortcuts } = this.props;
+
+		// Only render shortcuts if displayShortcuts is true
+		const shortcuts = displayShortcuts ? this.displayShortcutsComponent() : null;
```
Comment (getdave) [inline client/components/date-range/index.js]: Is this necessary? You have the `displayShortcuts` prop so you can conditionalise the rendering below based on that.

File: client/components/date-range/index.js [hunk 8]
```diff
@@ -461,26 +469,99 @@ export class DateRange extends Component {
...
-						selectedEndDate={ this.state.endDate }
-						onDateRangeChange={ onDateRangeChange }
-						focusedMonth={ this.state.focusedMonth }
-						numberOfMonths={ this.getNumberOfMonths() }
-					/>
+					{ shortcuts && (
+						<div className="date-control-picker-shortcuts">
+							{ this.displayShortcutsComponent() }
+						</div>
+					) }
```
Comment (getdave) [inline client/components/date-range/index.js]:
```suggestion
					{ displayShortcuts && (
						<div className="date-control-picker-shortcuts">
							{ this.displayShortcutsComponent() }
						</div>
					) }
```

Do you need the additional `shortcuts` variable?

**Review by getdave — COMMENTED** [review]
Some things I noticed on quick review.

File: client/components/date-range/index.js (hunk 7 above)
Comment (kangzj) [reply client/components/date-range/index.js]: Addressed in 519fd56ebf

File: client/components/date-range/shortcuts.tsx (hunk 6 above)
Comment (kangzj) [reply client/components/date-range/shortcuts.tsx]: I guess it was some interim status tweaking the code. Is it @annacmc 

I remove the code block in 766fe31ba0.

File: client/components/date-range/shortcuts.tsx (hunk 5 above)
Comment (kangzj) [reply client/components/date-range/shortcuts.tsx:76]: translated in getShortcutList already 😄

File: client/components/date-range/shortcuts.tsx (hunk 4 above)
Comment (kangzj) [reply client/components/date-range/shortcuts.tsx]: Addressed in af9e53beb1

**Review by kangzj — APPROVED** [review]
Thanks for the feedback @getdave all make sense 🙇 

Hi @annacmc I addressed all the feedback and remove some code that was resulted from the rebase, so that we could move on with iterating. Hope you don't mind. 

Let's 🚀

File: client/components/date-range/index.js [hunk 9]
```diff
@@ -387,6 +390,16 @@ export class DateRange extends Component {
...
 	}
 
+	handleDateRangeChange( startDate, endDate ) {
+		this.setState( {
+			startDate,
+			endDate,
+			textInputStartDate: this.toDateString( startDate ),
+			textInputEndDate: this.toDateString( endDate ),
+		} );
+		this.props.onDateSelect && this.props.onDateSelect( startDate, endDate );
```
Comment (kangzj) [inline client/components/date-range/index.js:400]: Not related, but I'll have to look at how this works 🤔
